src/data/worker.ts

```diff
@@ -0,0 +1,71 @@
+import { expose } from "comlink";
+import { parse } from "papaparse";
+import { CSV, Table } from "./";
+
+/**
+ * every time you communicate with a web worker, the message content must be
+ * serialized/deserialized, which takes a long time with large data. therefore,
+ * keep as much computation in this file as possible, and prune data as much as
+ * possible before returning.
+ */
+
+/** fetch file and parse as csv */
+export const parseData = async (url: string): Promise<CSV> => {
+  progress?.("Fetching");
+
+  const headers = new Headers();
+  headers.set("Accept-Encoding", "gzip, deflate, br");
+
+  const response = await fetch(import.meta.env.BASE_URL + url, { headers });
+  if (!response.ok) throw Error("Response not OK");
+
+  progress?.("Parsing text");
+  const text = await response.text();
+
+  progress?.("Parsing csv");
+  const parsed = await parse(text.trim());
+  const data = parsed.data as CSV;
+
+  return data;
+};
+
+/** parse csv with "by table" format */
+export const parseTable = async (url: string): Promise<Table> => {
+  const data = await parseData(url);
+
+  progress?.("Parsing table");
+
+  const table: Table = [];
+
+  for (let col = 1; col < data[0].length; col++) {
+    const fullName = String(data[0][col] || "");
+    /** get parts from full name */
+    const [kingdom = "", phylum = "", _class = ""] = fullName.split(".");
+    /** get name from most specific part */
+    const name = _class || phylum || kingdom;
+
+    /** count number of non-zero rows in col */
+    let samples = 0;
+    for (let row = 1; row < data.length; row++)
+      if (data[row][col] !== "0") samples++;
+
+    /** number */
+    if (name !== "NA")
```

Comment:
Would it be quicker to check this before looping through all the rows? If it's "NA," we can just continue to the next column, right?

Reply:
Indeed. I've moved it above the totaling step, but still after the `name` step. I figure we want to only skip it if the most specific level is exactly equal to `NA`?

Reply:
I was trying to comment on that line and then my internet died. That seems like a fine solution, but I think the name to be checked could be
`const name = _class || phylum;`
and exclude kingdom. It's not a big deal, but if the closest guess we can make is "it's more likely bacteria than archaea," that's probably a very flimsy guess.

Reply:
We're controlling this data, so hopefully one of those should always be defined? If we do that suggestion, what should be the fallback at the end, `""`? `"NA"` and skip?